// jsreport-studio/src/helpers/preview.js
import isObject from 'lodash/isObject'
import parseStreamingMultipart from './parseStreamingMultipart'
import resolveUrl from '../helpers/resolveUrl.js'

function addInput (form, name, value) {
  const input = document.createElement('input')
  input.type = 'hidden'
  input.name = name
  input.value = value
  form.appendChild(input)
}

export default async function (request, target) {
  delete request.template._id
  request.template.content = request.template.content || ' '

  request.options = request.options || {}
  request.options.preview = true

  if (target.type === 'download') {
    delete request.options.preview
    request.options.download = true
  }

  const templateName = request.template.name

  // NOTE: NEW implementation based on fetch
  let url = templateName ? resolveUrl(`/api/report/${encodeURIComponent(templateName)}`) : resolveUrl('/api/report')

  url = `${url}?logsStreaming=true`

  try {
    const response = await fetch(url, {
      method: 'POST',
      cache: 'no-cache',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        template: request.template,
        options: request.options,
        data: request.data
      })
    })

    let contentType = ''

    if (response.headers != null) {
      contentType = response.headers.get('Content-Type') || ''
    }

    if (response.status === 200) {
      if (contentType.indexOf('multipart/mixed') === -1) {
        throw new Error('Got invalid response content-type, expected multipart/mixed')
      }

      await parseStreamingMultipart(response, (fileInfo) => {
        if (fileInfo.name !== 'report') {
          console.log('name:', fileInfo.name, 'content:', new TextDecoder().decode(fileInfo.rawData))
        } else {
          // TODO: verify if using blob vs filename makes some difference in the preview of
          // files like pdf, probably when using file there is more metada because of the filename
          // but who knows..

          // const file = new Blob([fileInfo.rawData], { type: fileInfo.contentType })
          const file = new File([fileInfo.rawData], fileInfo.filename || '', {
            type: fileInfo.contentType
          })

          const localFileSrc = window.URL.createObjectURL(file)

          target.changeSrc(localFileSrc)
        }
      })
    } else {
      let content

      if (contentType === 'application/json') {
        content = await response.json()
      } else {
        content = await response.text()
      }

      const notOkError = new Error('Got not ok response')

      notOkError.data = content

      throw notOkError
    }
  } catch (e) {
    const newError = new Error(`Preview failed. ${e.message}`)

    newError.stack = e.stack
    Object.assign(newError, e)

    // throw newError
    console.error(newError)
  }

  // NOTE: OLD implementation based on FORMS
  /*
  const mapForm = document.createElement('form')

  mapForm.target = target
  mapForm.method = 'POST'

  // we set the template name in url just to show a title in the preview iframe, the name
  // won't be using at all on server side logic
  mapForm.action = templateName ? resolveUrl(`/api/report/${encodeURIComponent(templateName)}`) : resolveUrl('/api/report')

  function addBody (path, body) {
    if (body === undefined) {
      return
    }

    for (const key in body) {
      if (isObject(body[ key ])) {
        // if it is an empty object or array then it should not be added to form,
        // this fix problem with url encoded data which can not represent empty arrays or objects
        // so instead of sending empty `template[scripts]:` we don't add the value at all
        if (Object.keys(body[ key ]).length === 0) {
          continue
        }

        addBody(path + '[' + key + ']', body[ key ])
      } else {
        if (body[ key ] !== undefined && !(body[ key ] instanceof Array)) {
          addInput(mapForm, path + '[' + key + ']', body[ key ])
        }
      }
    }
  }

  addBody('template', request.template)
  addBody('options', request.options)

  if (request.data) {
    addInput(mapForm, 'data', request.data)
  }

  document.body.appendChild(mapForm)

  mapForm.submit()
  */
}
